File: src/main/java/org/jetbrains/plugins/featuresSuggester/PluginActivator.java
package org.jetbrains.plugins.featuresSuggester;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.fileEditor.FileEditorManagerListener;
import com.intellij.openapi.project.DumbService;
import com.intellij.openapi.util.registry.Registry;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;
import training.featuresSuggester.settings.FeatureSuggesterSettings;
import training.featuresSuggester.suggesters.FeatureSuggester;

import java.util.List;

public class PluginActivator implements FileEditorManagerListener {
    private static final String ENABLED_KEY = "feature.suggester.enable.suggesters";

    @Override
    public void fileOpened(@NotNull FileEditorManager source, @NotNull VirtualFile file) {
        ApplicationManager.getApplication().invokeLater(() -> {
            var project = source.getProject();
            if (project.isDisposed()) return;
            if (!Registry.is(ENABLED_KEY, true)) {
                var dumbService = DumbService.getInstance(project);
                if (dumbService.isDumb()) {
                    dumbService.runWhenSmart(this::enableSuggesters);
                } else {
                    enableSuggesters();
                }
            }
        });
    }

    private void enableSuggesters() {
        Registry.get(ENABLED_KEY).setValue(true);

        var settings = FeatureSuggesterSettings.instance();
        List<FeatureSuggester> suggesters = FeatureSuggester.Companion.getSuggesters();
        for (var suggester : suggesters) {
            settings.setEnabled(suggester.getId(), true);
        }
    }
}

File: src/main/kotlin/org/jetbrains/plugins/featuresSuggester/PluginActivator.kt
package org.jetbrains.plugins.featuresSuggester

import com.intellij.openapi.application.runInEdt
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.FileEditorManagerListener
import com.intellij.openapi.project.DumbService
import com.intellij.openapi.util.registry.Registry
import com.intellij.openapi.vfs.VirtualFile
import training.featuresSuggester.settings.FeatureSuggesterSettings
import training.featuresSuggester.suggesters.FeatureSuggester

class PluginActivator : FileEditorManagerListener {
    private val enabledKey = "feature.suggester.enable.suggesters"

    override fun fileOpened(source: FileEditorManager, file: VirtualFile) {
        runInEdt {
            val project = source.project
            if (project.isDisposed) return@runInEdt
            if (!Registry.`is`(enabledKey, true)) {
                val dumbService = DumbService.getInstance(project)
                if (dumbService.isDumb) {
                    dumbService.runWhenSmart(this::enableSuggesters)
                } else {
                    enableSuggesters()
                }
            }
        }
    }

    private fun enableSuggesters() {
        val settings = FeatureSuggesterSettings.instance()
        Registry.get(enabledKey).setValue(true)
        FeatureSuggester.suggesters.forEach {
            settings.setEnabled(it.id, true)
        }
    }
}